### PR TITLE
Feature/liayoo/client api access control

### PR DESCRIPTION
**Pull request by liayoo**
- Introduced DEV_CLIENT_API_IP_WHITELIST for client api access control.
- Added IP whitelisting json-rpc apis (ain_getDevClientApiIpWhitelist, ain_addToDevClientApiIpWhitelist, ain_removeFromDevClientApiIpWhitelist)
- Added IP whitelisting tool scripts
- Reorged json rpc apis
- Renamed ENABLE_DEV_CLIENT_SET_API -> ENABLE_DEV_CLIENT_API and updated to manage the get apis with the flag as well

### CODE REVIEW

File: json_rpc/index.js [hunk 1]
```diff
@@ -30,656 +26,27 @@ const PathUtil = require('../common/path-util');
...
-        const latency = Date.now() - beginTime;
-        trafficStatsManager.addEvent(TrafficEventTypes.JSON_RPC_SET, latency);
-        done(null, addProtocolVersion({ result }));
-      }
-    }
+    ...getStateModifyingApis(node, p2pServer),
   };
 
-  let methods = nonTxMethods;
+  const methods = nonTxMethods;
```
Comment (platfowner) [inline json_rpc/index.js:43]: Could you leave only minimum APIs (such as p2p_getPeerCandidateInfo()) outside of ENABLE_JSON_RPC_TX_API flag guarding and moving other APIs inside like in the case of the dev client API for me?

Comment (platfowner) [reply json_rpc/index.js:43]: maybe minimumMethods and nonMinimumMethods?

File: json_rpc/index.js [hunk 2]
```diff
@@ -30,656 +26,27 @@ const PathUtil = require('../common/path-util');
...
-        const result = p2pServer.executeAndBroadcastTransaction({tx_list: txList});
-        const latency = Date.now() - beginTime;
-        trafficStatsManager.addEvent(TrafficEventTypes.JSON_RPC_SET, latency);
-        done(null, addProtocolVersion({ result }));
-      }
-    }
+    ...getStateModifyingApis(node, p2pServer),
   };
 
-  let methods = nonTxMethods;
```
Comment (platfowner) [inline json_rpc/index.js:653]: thanks for fixing this!

File: client/index.js [hunk 3]
```diff
@@ -107,234 +109,252 @@ app.get('/metrics', (req, res, next) => {
...
-  trafficStatsManager.addEvent(TrafficEventTypes.CLIENT_API_GET, latency);
-  res.status(200)
-    .set('Content-Type', 'application/json')
-    .send({
-      code: result !== null ? DevClientApiResultCode.SUCCESS : DevClientApiResultCode.FAILURE,
-      result
-    })
-    .end();
-});
+app.use(ipWhitelist((ip) => {
```
Comment (platfowner) [inline client/index.js:124]: nice approach!

double-checking: is this setting applied only for the apis added below (not above)?

File: json_rpc/index.js (hunk 1 above)
Comment (liayoo) [reply json_rpc/index.js]: Sounds great! Thanks for the idea. Done.

File: client/index.js (hunk 3 above)
Comment (liayoo) [reply client/index.js:124]: Thanks! That's correct!

Comment (minsulee2) [reply client/index.js:124]: Wooooooow, Good to see this!

File: json_rpc/api-access.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+} = require('../common/constants');
+const { JsonRpcApiResultCode } = require('../common/result-code');
+const CommonUtil = require('../common/common-util');
+const JsonRpcUtil = require('./util');
+
+module.exports = function getApiAccessApis(node) {
+  return {
+    ain_getDevClientApiIpWhitelist: function(args, done) {
+      const beginTime = Date.now();
+      const verified = node.verifyNodeAccountSignature(args.message, args.signature);
```
Comment (minsulee2) [inline json_rpc/api-access.js:16]: What about checking sig first then do the last of them later?

File: json_rpc/api-access.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        done(null, JsonRpcUtil.addProtocolVersion({ result: NodeConfigs.DEV_CLIENT_API_IP_WHITELIST }));
+      } else {
+        done({ code: 403, message: 'Forbidden' });
+      }
+    },
+
+    ain_addToDevClientApiIpWhitelist: function(args, done) {
+      const beginTime = Date.now();
+      const verified = node.verifyNodeAccountSignature(args.message, args.signature);
+      if (_.get(args.message, 'method') !== 'ain_addToDevClientApiIpWhitelist' || !verified) {
```
Comment (minsulee2) [inline json_rpc/api-access.js:29]: ditto

Comment (minsulee2) [reply json_rpc/api-access.js:29]: This seems ok tho. sorry for confusion!

File: json_rpc/api-access.js (hunk 4 above)
Comment (minsulee2) [reply json_rpc/api-access.js:16]: As discussed offline, never mind this comment!

File: json_rpc/account.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    ain_getTimestamp: function(args, done) {
+      const beginTime = Date.now();
+      const result = node.getTimestampForAddr(args.address, args.from === 'pending');
+      const latency = Date.now() - beginTime;
+      trafficStatsManager.addEvent(TrafficEventTypes.JSON_RPC_GET, latency);
+      done(null, JsonRpcUtil.addProtocolVersion({ result }));
+    },
+
+    ain_isValidator: function(args, done) {
```
Comment (platfowner) [inline json_rpc/account.js:87]: (seems to be out of scope but..) what about 'ain_getValidatorStake'?
or returning { isValidator: true/false, stake: <stake amount> } instead?

File: json_rpc/index.js [hunk 7]
```diff
@@ -27,659 +22,28 @@ const PathUtil = require('../common/path-util');
...
+    Object.assign(apis, {
+      ...getBlockApis(node),
+      ...getDatabaseApis(node),
+      ...getNetworkApis(node, p2pServer),
+      ...getTransactionApis(node, p2pServer),
+      ...getVersionApis(minProtocolVersion, maxProtocolVersion),
+    });
+  } else {
+    Object.assign(apis, {
+      p2p_getPeerCandidateInfo: getNetworkApis(node, p2pServer).p2p_getPeerCandidateInfo,
```
Comment (platfowner) [inline json_rpc/index.js:41]: 👍

File: json_rpc/util.js [hunk 8]
```diff
@@ -0,0 +1,22 @@
+
```
Comment (platfowner) [inline json_rpc/util.js:1]: minor: what about naming json-rpc-util.js like others (e.g. network-util.js)?

File: tools/api-access/util.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  // NOTE(liayoo): Show the prompt & hide the password
+  readlineInterface._writeToOutput = (val) => {
+    if (!hide && secret) {
+      readlineInterface.output.write(val);
+      hide = true;
+    } else if (!hide && !secret) {
+      readlineInterface.output.write(val);
+    }
+  };
+  switch (type) {
```
Comment (platfowner) [inline tools/api-access/util.js:23]: 👍

**Review by platfowner — APPROVED** [review]
LGTM with some minor comments.
Thanks for the nice implementation!

**Review by minsulee2 — APPROVED** [review]
LGTM! Thanks for the long explains! I got this!

File: blockchain-configs/afan-shard/node_params.json [hunk 10]
```diff
@@ -16,11 +16,12 @@
     "http://localhost:3000"
   ],
   "DEBUG": false,
-  "ENABLE_DEV_CLIENT_SET_API": false,
+  "DEV_CLIENT_API_IP_WHITELIST": "*",
+  "ENABLE_DEV_CLIENT_API": true,
```
Comment (platfowner) [inline blockchain-configs/afan-shard/node_params.json:20]: as discussed offline, using this flag for 'set' APIs only, while controlling others with the whitelist flag value?

File: blockchain-configs/afan-shard/node_params.json [hunk 11]
```diff
@@ -16,11 +16,12 @@
     "http://localhost:3000"
   ],
   "DEBUG": false,
-  "ENABLE_DEV_CLIENT_SET_API": false,
+  "DEV_CLIENT_API_IP_WHITELIST": "*",
+  "ENABLE_DEV_CLIENT_SET_API": true,
```
Comment (platfowner) [inline blockchain-configs/afan-shard/node_params.json:20]: minor) would it be just safer to keep these values 'false' in gcp configs?
it seems the corresponding 'set' apis are used in tests only.

Comment (liayoo) [reply blockchain-configs/afan-shard/node_params.json:20]: right! updated. thanks~

File: blockchain-configs/afan-shard/node_params.json (hunk 10 above)
Comment (liayoo) [reply blockchain-configs/afan-shard/node_params.json]: yes! updated!

File: json_rpc/util.js (hunk 8 above)
Comment (liayoo) [reply json_rpc/util.js]: done~

File: json_rpc/account.js (hunk 6 above)
Comment (liayoo) [reply json_rpc/account.js]: Added ain_getValidatorInfo that returns { isWhitelisted, stake }. PTAL~

Comment (platfowner) [reply json_rpc/account.js]: looks good, thanks!

**Comment by liayoo** [comment]
Thanks for the reviews! Now merging~

**Pull request merged**